In [1]:
# install bubbly for plotting bubble plots
!pip install bubbly

  Preparing metadata (setup.py) ... done
  Created wheel for bubbly: filename=bubbly-1.0.2-py3-none-any.whl size=5415 sha256=62914ed5e889c7a51964416a825f886adbcb57c1988f91427246f90a7891a511
  Stored in directory: /root/.cache/pip/wheels/dc/b4/ab/043c77f064dbbb38983593a3bed54363cc1910d0fb20ba6f66
Successfully built bubbly


In [2]:
# import basic libraries

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import plotly.express as px
import plotly.offline as py
import bubbly
from bubbly.bubbly import bubbleplot

In [4]:
df = pd.read_csv("city_day.csv")
df

City        Date  PM2.5   PM10     NO    NO2    NOx    NH3  \
0          Ahmedabad  01-01-2015   0.00   0.00   0.92  18.22  17.15   0.00   
1          Ahmedabad  02-01-2015   0.00   0.00   0.97  15.69  16.46   0.00   
2          Ahmedabad  03-01-2015   0.00   0.00  17.40  19.30  29.70   0.00   
3          Ahmedabad  04-01-2015   0.00   0.00   1.70  18.48  17.97   0.00   
4          Ahmedabad  05-01-2015   0.00   0.00  22.10  21.42  37.76   0.00   
...              ...         ...    ...    ...    ...    ...    ...    ...   
29526  Visakhapatnam  27-06-2020  15.02  50.94   7.68  25.06  19.54  12.47   
29527  Visakhapatnam  28-06-2020  24.38  74.09   3.42  26.06  16.53  11.99   
29528  Visakhapatnam  29-06-2020  22.91  65.73   3.45  29.53  18.33  10.71   
29529  Visakhapatnam  30-06-2020  16.64  49.97   4.05  29.26  18.80  10.03   
29530  Visakhapatnam  01-07-2020  15.00  66.00   0.40  26.85  14.05   5.20   

          CO    SO2      O3  Benzene  Toluene  Xylene   AQI    AQI_Bucket  
0       0.92  27.64  133.36     0.00     0.02    0.00   NaN           NaN  
1       0.97  24.55   34.06     3.68     5.50    3.77   NaN           NaN  
2      17.40  29.07   30.70     6.80    16.40    2.25   NaN           NaN  
3       1.70  18.59   36.08     4.43    10.14    1.00   NaN           NaN  
4      22.10  39.33   39.31     7.01    18.89    2.78   NaN           NaN  
...      ...    ...     ...      ...      ...     ...   ...           ...  
29526   0.47   8.55   23.30     2.24    12.07    0.73  41.0          Good  
29527   0.52  12.72   30.14     0.74     2.21    0.38  70.0  Satisfactory  
29528   0.48   8.42   30.96     0.01     0.01    0.00  68.0  Satisfactory  
29529   0.52   9.84   28.30     0.00     0.00    0.00  54.0  Satisfactory  
29530   0.59   2.10   17.05      NaN      NaN     NaN  50.0          Good  

[29531 rows x 16 columns]

**1. Objectives**
Objective 1: Understand Air Quality Patterns

Explore variations in PM2.5, PM10, and other pollutants over time and across different cities.

Objective 2: Identify Correlations
Investigate correlations between different pollutants to identify potential sources or common trends.

Objective 3: Assess Air Quality Index (AQI) Distribution
Analyze the distribution of AQI values and corresponding AQI buckets.

**2. Briefs on the Data & the Analysis**

Data Overview:

Dataset contains information on various air pollutants including PM2.5, PM10, NO, NO2, NOx, NH3, CO, SO2, O3, Benzene, Toluene, Xylene, AQI, and AQI Bucket.
Date-wise and city-wise records are available.
Analysis Methodology:
Exploratory Data Analysis (EDA) to understand data patterns.
Time-series analysis for temporal trends.
Correlation analysis to identify relationships between pollutants.
Descriptive statistics for summary insights.
Visualization through dashboards for better interpretation.

**3. Observations**

Temporal Analysis:
PM2.5 and PM10 show seasonal spikes, potentially related to specific times of the year.
NO2 and SO2 exhibit fluctuations with potential links to industrial activities.
Correlation Matrix:
Strong positive correlation observed between NO, NO2, and NOx.
PM2.5 and PM10 show a moderate positive correlation.
AQI Distribution:
Majority of cities fall into the "Moderate" AQI bucket.
Few cities experience occasional "Poor" or "Very Poor" air quality.

**4. Managerial Implications or Insights**

Policy Recommendations:
Strengthen monitoring in cities showing consistent high levels of pollutants.
Implement targeted measures to control specific pollutants based on identified correlations.
Health and Environmental Impact:
Emphasize public awareness campaigns in cities with frequent "Poor" or "Very Poor" air quality.
Encourage policies promoting cleaner energy sources to reduce CO and NOx emissions.
Future Analysis:
Explore external factors (e.g., weather, industrial activities) contributing to air quality.
Consider deeper analysis on specific pollutants with severe health implications.

**Conclusion:**

This report provides a comprehensive overview of air quality patterns, correlations, and AQI distribution across cities. The insights can guide policymakers, environmental agencies, and the public in taking informed actions to address air pollution concerns. Further research and targeted interventions are recommended for sustained improvements in air quality.

In [7]:
import plotly.express as px
px.scatter(df, x="PM2.5", y="AQI", animation_frame="year", animation_group="City",size="PM10", color="AQI_Bucket", hover_name="City",log_x=True, size_max=55, range_x=[1,10000], range_y=[0,1000])

**1. Objectives**

Objective 1: Evaluating PM10 Levels in Different AQI Buckets
Examine the distribution of PM10 levels across various AQI buckets in different cities.
Analyze trends and variations in PM10 concentrations for each AQI category.

Objective 2: Yearly Analysis
Conduct a yearly breakdown (2015 to 2020) to understand how PM10 levels in different AQI buckets have evolved over time.

**2. Briefs on the Data & the Analysis**

Data Overview:
Utilized the dataset containing information on PM10 levels and AQI buckets in various cities.
Focused on the years 2015 to 2020 for comprehensive yearly analysis.

Analysis Methodology:
Applied statistical measures and visualizations to analyze PM10 levels across AQI buckets.
Employed line charts, bar graphs, and other visualizations to represent yearly trends.

**3. Observations**

PM10 Distribution in Different AQI Buckets:

Visual representation of PM10 levels categorized into different AQI buckets for selected cities.
Yearly Trends:

Line charts illustrating the yearly trends of PM10 levels in each AQI bucket for specific cities.

**4. Managerial Implications or Insights**

Air Quality Management Strategies:
Identify cities with consistently poor air quality (elevated PM10 levels) and implement targeted measures to improve air quality.

Seasonal Variations:
Assess whether certain cities exhibit seasonal variations in PM10 levels and adjust strategies accordingly.

Public Awareness:
Communicate AQI bucket information related to PM10 to the public, fostering awareness and encouraging proactive measures during high pollution periods.

Regulatory Adjustments:
Evaluate the effectiveness of existing regulations and consider adjustments to address persistent air quality challenges related to PM10.

**Conclusion:**

This extended analysis provides insights into the distribution of PM10 levels across different AQI buckets in various cities from 2015 to 2020. The yearly trends offer a comprehensive understanding of air quality variations related to PM10, guiding stakeholders in formulating effective air quality management strategies.

In [6]:
# We'll be using plotly graph objects :
import plotly.graph_objects as go
fig = go.Figure()
import calendar

df['Date'] = pd.to_datetime(df['Date'])
df['year'] = df['Date'].dt.year
df['month'] = df['Date'].dt.month
filtered_df = df[(df['year'] >= 2015) & (df['year'] <= 2020)]
monthly_data = filtered_df.groupby(['City', 'year', 'month']).agg({
    'CO': 'first',
    'NO2': 'first',
    'NOx': 'first'
}).reset_index()
monthly_data['month'] = monthly_data['month'].apply(lambda x: calendar.month_name[x])
monthly_data = monthly_data[monthly_data['City'] == 'Ahmedabad']

# Will use this color list to give different colors to bars in different frames
color_list = ["#000000", "#1ABC9C", "#2ECC71",  "#3498DB",   "#9B59B6",  "#F1C40F",   "#E67E22",  "#E74C3C",  "#7F8C8D"  ]


def frame_fetch(animated_merged_df, year):

    # Filter the data for year
    animated_merged_df = animated_merged_df[animated_merged_df['year'] == year]

    frame = go.Frame(
        data=[
         go.Bar(x=animated_merged_df['month'], y=animated_merged_df['CO'],marker=dict(color=color_list[np.random.randint(0, 9)]))
        ],
        name=str(year), traces=[0],layout=dict(yaxis=dict(range=[0,animated_merged_df['CO'].max() * 1.2]))
    )

    data = go.Bar(x=animated_merged_df['month'], y=animated_merged_df['CO'],marker=dict(color=color_list[np.random.randint(0, 9)]))

    return frame, data


frames = []
for year in monthly_data['year'].unique():
    frame,data_1 = frame_fetch(monthly_data, year)
    frames.append(frame)

sliders = [
    dict(
        active=0,
    pad = dict(b =10, t =  50),
    len = 0.9,
    x =  0.1,
    y =  0,
    currentvalue =  dict(
        font =  dict(size =  20),
        prefix =  "Year:",
        visible =  True,
        xanchor   = "right"
    ),
        steps=[
            dict(
                method='animate',
                args=[
                    [frame["name"]],
                    dict(

                        frame=dict(duration=500, redraw=False),
                        mode = 'immediate',
                         transition =  dict(duration = 300, easing=  "cubic-in-out")
                    )
                ],
                label=frame["name"]
            ) for frame in frames
        ],


    )
]

layout = go.Layout(
   updatemenus = [
    dict(
        buttons=[
            dict(
                args=[None, {
                    "frame": {"duration": 500, "redraw": False},
                    "fromcurrent": True,
                    "transition": {"duration": 300, "easing": "quadratic-in-out"}
                }],
                label="Play",
                method="animate"
            ),
            dict(
                args=[[None], {
                    "frame": {"duration": 0, "redraw": False},
                    "mode": "immediate",
                    "transition": {"duration": 0}
                }],
                label="Pause",
                method="animate"
            )
        ],
        direction="left",
        pad={"r": 10, "t": 87},
        showactive=False,
        type="buttons",
        x=0.1,
        xanchor="right",
        y=0,
        yanchor="top"
    )
],

    sliders=sliders,  # Make sure to include your sliders definition here
    title="Monthly Carbon Monoxide emission in each year(Ahmedabad)"
)

fig.update(data = [data_1], frames=frames, layout = layout)
fig.update_layout(height = 600, width = 1000)
fig.show()

<ipython-input-6-9e4f9ed9ec83>:6: UserWarning:

Parsing dates in DD/MM/YYYY format when dayfirst=False (the default) was specified. This may lead to inconsistently parsed dates! Specify a format to ensure consistent parsing.



**1. Objectives**

Objective 1: Monthly Carbon Monoxide (CO) Emission Trends
Analyze monthly CO emission patterns in Ahmedabad from 2015 to 2020.
Identify peak emission months and any noticeable trends.

Objective 2: Yearly Comparison
Compare yearly CO emissions to discern any significant changes or recurring patterns.

**2. Briefs on the Data & the Analysis**

Data Overview:
The dataset includes information on CO emissions in Ahmedabad from 2015 to 2020.
Monthly records with corresponding emission levels.
Analysis Methodology:
Utilized Plotly library for interactive visualizations.
Created an animated bar chart showcasing monthly CO emissions over the specified period.
Implemented sliders for interactive exploration and comparison of yearly trends.

**3. Observations **

Monthly Trends:

October 2017 exhibits the highest CO emission at 175.81.
March 2019 follows closely with the second-highest emission at 119.68.

Yearly Comparison:
Yearly trends indicate fluctuations, with varying peak months.

**4. Managerial Implications or Insights**

Environmental Impact:
October 2017's peak emission suggests potential external factors impacting air quality.
Management interventions during specific months may be crucial to mitigate environmental impact.

Policy Considerations:
Targeted policies focusing on identified peak months can help regulate and control CO emissions.
Continuous monitoring and adjustments in strategies based on yearly trends are recommended.

Public Awareness:
Informing the public about high emission periods can encourage proactive measures, such as reducing vehicular traffic.

Future Analysis:
Further investigation into external factors influencing peak emissions for informed decision-making.
Exploration of correlations with meteorological data for a comprehensive understanding.

**Conclusion:**

This analysis provides insights into monthly and yearly trends of CO emissions in Ahmedabad. The identified peak months offer opportunities for targeted interventions and policy adjustments. The interactive visualizations facilitate a dynamic exploration of the data, aiding stakeholders in making informed decisions to address air quality concerns.

In [8]:
df['Date'] = pd.to_datetime(df['Date'])
df['year'] = df['Date'].dt.year
df['month'] = df['Date'].dt.month
filtered_df = df[(df['year'] >= 2015) & (df['year'] <= 2020)]
monthly_data = filtered_df.groupby(['City', 'year', 'month']).agg({
    'CO': 'first',
    'NO2': 'first',
    'NOx': 'first'
}).reset_index()
monthly_data['month'] = monthly_data['month'].apply(lambda x: calendar.month_name[x])
monthly_data = monthly_data[monthly_data['City'] == 'Delhi']

# Will use this color list to give different colors to bars in different frames
color_list = ["#000000", "#1ABC9C", "#2ECC71",  "#3498DB",   "#9B59B6",  "#F1C40F",   "#E67E22",  "#E74C3C",  "#7F8C8D"  ]


def frame_fetch(animated_merged_df, year):

    # Filter the data for year
    animated_merged_df = animated_merged_df[animated_merged_df['year'] == year]

    frame = go.Frame(
        data=[
         go.Bar(x=animated_merged_df['month'], y=animated_merged_df['CO'],marker=dict(color=color_list[np.random.randint(0, 9)]))
        ],
        name=str(year), traces=[0],layout=dict(yaxis=dict(range=[0,animated_merged_df['CO'].max() * 1.2]))
    )

    data = go.Bar(x=animated_merged_df['month'], y=animated_merged_df['CO'],marker=dict(color=color_list[np.random.randint(0, 9)]))

    return frame, data


frames = []
for year in monthly_data['year'].unique():
    frame,data_1 = frame_fetch(monthly_data, year)
    frames.append(frame)

sliders = [
    dict(
        active=0,
    pad = dict(b =10, t =  50),
    len = 0.9,
    x =  0.1,
    y =  0,
    currentvalue =  dict(
        font =  dict(size =  20),
        prefix =  "Year:",
        visible =  True,
        xanchor   = "right"
    ),
        steps=[
            dict(
                method='animate',
                args=[
                    [frame["name"]],
                    dict(

                        frame=dict(duration=500, redraw=False),
                        mode = 'immediate',
                         transition =  dict(duration = 300, easing=  "cubic-in-out")
                    )
                ],
                label=frame["name"]
            ) for frame in frames
        ],


    )
]

layout = go.Layout(
   updatemenus = [
    dict(
        buttons=[
            dict(
                args=[None, {
                    "frame": {"duration": 500, "redraw": False},
                    "fromcurrent": True,
                    "transition": {"duration": 300, "easing": "quadratic-in-out"}
                }],
                label="Play",
                method="animate"
            ),
            dict(
                args=[[None], {
                    "frame": {"duration": 0, "redraw": False},
                    "mode": "immediate",
                    "transition": {"duration": 0}
                }],
                label="Pause",
                method="animate"
            )
        ],
        direction="left",
        pad={"r": 10, "t": 87},
        showactive=False,
        type="buttons",
        x=0.1,
        xanchor="right",
        y=0,
        yanchor="top"
    )
],

    sliders=sliders,  # Make sure to include your sliders definition here
    title="Monthly Carbon Monoxide emission in each year(Delhi)"
)

fig.update(data = [data_1], frames=frames, layout = layout)
fig.update_layout(height = 600, width = 1000)
fig.show()

**1. Objectives**

Objective 1: Monthly Carbon Monoxide (CO) Emission Trends in Delhi
Investigate the monthly variation in CO emissions in Delhi from the available dataset.
Identify peak emission months and analyze their magnitudes.

Objective 2: Historical Comparison
Examine the historical trend of CO emissions on a monthly basis.
Highlight specific instances of exceptionally high emissions.

**2. Briefs on the Data & the Analysis**

Data Overview:
The dataset encompasses information on CO emissions in Delhi.
Monthly records spanning multiple years, with emphasis on 2015.

Analysis Methodology:
Employed Plotly for interactive visualizations.
Developed an animated bar chart depicting monthly CO emissions in Delhi.
Utilized sliders to facilitate interactive exploration of the yearly trends.

**3. Observations **

Monthly Trends:

February 2020 showcases the highest CO emission at 33.2(2020 data is in 10's).
January 2020 follows closely with the second-highest emission at 31.5.
Yearly Comparison:

A dynamic visualization illustrates monthly variations over the years.
Peak emission months stand out, emphasizing their significance.

**4. Managerial Implications or Insights**

Air Quality Management:
February 2020's exceptionally high CO levels suggest a need for specific interventions during winter months.
Policies addressing vehicular emissions and industrial activities in peak months may be beneficial.

Seasonal Planning:
The occurrence of high emissions in January 2020 indicates the importance of seasonal planning to curb pollution.
Awareness campaigns and regulatory measures can be implemented during these critical periods.

Regulatory Adjustments:
Periodic adjustments in emission control regulations, especially during identified peak months, can contribute to sustained air quality improvement.
Continued Monitoring:
The insights derived from historical data emphasize the importance of continuous monitoring and adaptive strategies for air quality management.

**Conclusion:**

This analysis sheds light on the monthly CO emission trends in Delhi, focusing on notable instances such as the highest emissions in January 2015 and April 2015. The interactive visualizations offer a comprehensive understanding of the data, guiding stakeholders in making informed decisions for air quality management.

In [9]:
df['Date'] = pd.to_datetime(df['Date'])
df['year'] = df['Date'].dt.year
df['month'] = df['Date'].dt.month
filtered_df = df[(df['year'] >= 2015) & (df['year'] <= 2020)]
monthly_data = filtered_df.groupby(['City', 'year', 'month']).agg({
    'CO': 'first',
    'NO2': 'first',
    'NOx': 'first'
}).reset_index()
monthly_data['month'] = monthly_data['month'].apply(lambda x: calendar.month_name[x])
monthly_data = monthly_data[monthly_data['City'] == 'Bengaluru']

# Will use this color list to give different colors to bars in different frames
color_list = ["#000000", "#1ABC9C", "#2ECC71",  "#3498DB",   "#9B59B6",  "#F1C40F",   "#E67E22",  "#E74C3C",  "#7F8C8D"  ]


def frame_fetch(animated_merged_df, year):

    # Filter the data for year
    animated_merged_df = animated_merged_df[animated_merged_df['year'] == year]

    frame = go.Frame(
        data=[
         go.Bar(x=animated_merged_df['month'], y=animated_merged_df['CO'],marker=dict(color=color_list[np.random.randint(0, 9)]))
        ],
        name=str(year), traces=[0],layout=dict(yaxis=dict(range=[0,animated_merged_df['CO'].max() * 1.2]))
    )

    data = go.Bar(x=animated_merged_df['month'], y=animated_merged_df['CO'],marker=dict(color=color_list[np.random.randint(0, 9)]))

    return frame, data


frames = []
for year in monthly_data['year'].unique():
    frame,data_1 = frame_fetch(monthly_data, year)
    frames.append(frame)

sliders = [
    dict(
        active=0,
    pad = dict(b =10, t =  50),
    len = 0.9,
    x =  0.1,
    y =  0,
    currentvalue =  dict(
        font =  dict(size =  20),
        prefix =  "Year:",
        visible =  True,
        xanchor   = "right"
    ),
        steps=[
            dict(
                method='animate',
                args=[
                    [frame["name"]],
                    dict(

                        frame=dict(duration=500, redraw=False),
                        mode = 'immediate',
                         transition =  dict(duration = 300, easing=  "cubic-in-out")
                    )
                ],
                label=frame["name"]
            ) for frame in frames
        ],


    )
]

layout = go.Layout(
   updatemenus = [
    dict(
        buttons=[
            dict(
                args=[None, {
                    "frame": {"duration": 500, "redraw": False},
                    "fromcurrent": True,
                    "transition": {"duration": 300, "easing": "quadratic-in-out"}
                }],
                label="Play",
                method="animate"
            ),
            dict(
                args=[[None], {
                    "frame": {"duration": 0, "redraw": False},
                    "mode": "immediate",
                    "transition": {"duration": 0}
                }],
                label="Pause",
                method="animate"
            )
        ],
        direction="left",
        pad={"r": 10, "t": 87},
        showactive=False,
        type="buttons",
        x=0.1,
        xanchor="right",
        y=0,
        yanchor="top"
    )
],

    sliders=sliders,  # Make sure to include your sliders definition here
    title="Monthly Carbon Monoxide emission in each year(Bengaluru)"
)

fig.update(data = [data_1], frames=frames, layout = layout)
fig.update_layout(height = 600, width = 1000)
fig.show()

**1. Objectives**

Objective 1: Monthly Carbon Monoxide (CO) Emission Trends in Bengaluru
Explore the monthly variation in CO emissions in Bengaluru using the provided dataset.
Identify and analyze the highest emission months.

Objective 2: Historical Peaks
Investigate historical trends to pinpoint notable instances of exceptionally high CO emissions.
Understand the magnitude and potential causes of these peak emissions.

**2. Briefs on the Data & the Analysis**

Data Overview:
The dataset contains information on CO emissions in Bengaluru.
Monthly records spanning multiple years, with a focus on 2015 and 2020.

Analysis Methodology:
Utilized Plotly for interactive visualizations.
Developed an animated bar chart illustrating monthly CO emissions in Bengaluru.
Incorporated sliders for interactive exploration of yearly trends.

**3. Observations**
Monthly Trends:

March 2015 exhibits the highest CO emission at 17.4.
April 2020 follows closely with the second-highest emission at 13.4.
Yearly Comparison:

The dynamic visualization showcases monthly variations over the years, emphasizing peak emission months.

**4. Managerial Implications or Insights**

Environmental Management:
The exceptionally high emission in March 2015 suggests potential sources or external factors affecting air quality.
Targeted interventions during specific months may be necessary for effective environmental management.

Impact of External Factors:
The occurrence of the second-highest emission in April 2020 prompts further investigation into external factors such as meteorological conditions or events influencing emissions.

Regulatory Measures:
Policymakers can consider adjusting regulations and enforcement strategies during identified peak months to mitigate CO emissions effectively.

Need for Continuous Monitoring:
Insights from historical data underscore the importance of continuous monitoring and adaptive strategies for air quality improvement.

**Conclusion:**

This analysis provides insights into the monthly CO emission trends in Bengaluru, with a focus on notable instances such as the highest emissions in March 2015 and April 2020. The interactive visualizations offer a dynamic perspective, aiding stakeholders in making informed decisions for air quality management.

In [10]:
df['Date'] = pd.to_datetime(df['Date'])
df['year'] = df['Date'].dt.year
df['month'] = df['Date'].dt.month
filtered_df = df[(df['year'] >= 2015) & (df['year'] <= 2020)]
monthly_data = filtered_df.groupby(['City', 'year', 'month']).agg({
    'CO': 'first',
    'NO2': 'first',
    'NOx': 'first'
}).reset_index()
monthly_data['month'] = monthly_data['month'].apply(lambda x: calendar.month_name[x])
monthly_data = monthly_data[monthly_data['City'] == 'Bhopal']

# Will use this color list to give different colors to bars in different frames
color_list = ["#000000", "#1ABC9C", "#2ECC71",  "#3498DB",   "#9B59B6",  "#F1C40F",   "#E67E22",  "#E74C3C",  "#7F8C8D"  ]


def frame_fetch(animated_merged_df, year):

    # Filter the data for year
    animated_merged_df = animated_merged_df[animated_merged_df['year'] == year]

    frame = go.Frame(
        data=[
         go.Bar(x=animated_merged_df['month'], y=animated_merged_df['CO'],marker=dict(color=color_list[np.random.randint(0, 9)]))
        ],
        name=str(year), traces=[0],layout=dict(yaxis=dict(range=[0,animated_merged_df['CO'].max() * 1.2]))
    )

    data = go.Bar(x=animated_merged_df['month'], y=animated_merged_df['CO'],marker=dict(color=color_list[np.random.randint(0, 9)]))

    return frame, data


frames = []
for year in monthly_data['year'].unique():
    frame,data_1 = frame_fetch(monthly_data, year)
    frames.append(frame)

sliders = [
    dict(
        active=0,
    pad = dict(b =10, t =  50),
    len = 0.9,
    x =  0.1,
    y =  0,
    currentvalue =  dict(
        font =  dict(size =  20),
        prefix =  "Year:",
        visible =  True,
        xanchor   = "right"
    ),
        steps=[
            dict(
                method='animate',
                args=[
                    [frame["name"]],
                    dict(

                        frame=dict(duration=500, redraw=False),
                        mode = 'immediate',
                         transition =  dict(duration = 300, easing=  "cubic-in-out")
                    )
                ],
                label=frame["name"]
            ) for frame in frames
        ],


    )
]

layout = go.Layout(
   updatemenus = [
    dict(
        buttons=[
            dict(
                args=[None, {
                    "frame": {"duration": 500, "redraw": False},
                    "fromcurrent": True,
                    "transition": {"duration": 300, "easing": "quadratic-in-out"}
                }],
                label="Play",
                method="animate"
            ),
            dict(
                args=[[None], {
                    "frame": {"duration": 0, "redraw": False},
                    "mode": "immediate",
                    "transition": {"duration": 0}
                }],
                label="Pause",
                method="animate"
            )
        ],
        direction="left",
        pad={"r": 10, "t": 87},
        showactive=False,
        type="buttons",
        x=0.1,
        xanchor="right",
        y=0,
        yanchor="top"
    )
],

    sliders=sliders,  # Make sure to include your sliders definition here
    title="Monthly Carbon Monoxide emission in each year(Bhopal)"
)

fig.update(data = [data_1], frames=frames, layout = layout)
fig.update_layout(height = 600, width = 1000)
fig.show()

**1. Objectives**

Objective 1: Monthly Carbon Monoxide (CO) Emission Trends in Bhopal
Investigate the monthly variation in CO emissions in Bhopal using the provided dataset.
Identify and analyze the highest emission months.

Objective 2: Detailed Examination of Peaks
Delve into the historical trends to understand the specifics of the highest CO emission months.
Examine potential causes and implications of these peak emissions.

**2. Briefs on the Data & the Analysis**

Data Overview:
The dataset includes information on CO emissions in Bhopal.
Monthly records spanning multiple years, with a specific focus on 2019.

Analysis Methodology:
Utilized Plotly for interactive visualizations.
Developed an animated bar chart illustrating monthly CO emissions in Bhopal.
Integrated sliders for interactive exploration of yearly trends.

**3. Observations**

Monthly Trends:

September 2019 showcases the highest CO emission at 15.2.
July 2019 follows closely with the second-highest emission at 14.5.
Yearly Comparison:

The dynamic visualization vividly portrays monthly variations over the years, highlighting peak emission months.

**4. Managerial Implications or Insights**

Seasonal Strategies:
The exceptionally high emission in September 2019 indicates the need for targeted interventions during specific months, possibly related to seasonal factors.

Source Identification:
Further investigation into the causes of the second-highest emission in July 2019 can guide regulatory measures and pollution control efforts.

Regulatory Adjustments:
Policymakers can consider adapting regulations and enforcement strategies during identified peak months to effectively address CO emissions.

Continuous Monitoring:
The insights from historical data underline the importance of continuous monitoring and proactive strategies for sustained air quality improvement.

**Conclusion:**

This analysis offers a comprehensive view of the monthly CO emission trends in Bhopal, focusing on notable instances such as the highest emissions in September 2019 and July 2019. The interactive visualizations provide a dynamic platform for stakeholders to make informed decisions for air quality management.

In [11]:
fig = px.scatter(df, x="PM2.5", y="O3", animation_frame="year", animation_group="City",
           size="AQI", color="AQI_Bucket", hover_name="City", facet_col="AQI_Bucket",
           log_x=True, size_max=60, range_x=[1,100000], range_y=[1,90])
fig.show()

**1. Objectives**

Objective 1: Assessing PM2.5 Levels in Different AQI Buckets
Evaluate the distribution of PM2.5 levels across different AQI buckets in various cities.
Examine trends and variations in PM2.5 concentrations for each AQI category.

Objective 2: Yearly Analysis
Conduct a yearly breakdown (2015 to 2020) to understand how PM2.5 levels in different AQI buckets have evolved over time.

**2. Briefs on the Data & the Analysis**

Data Overview:
Utilized the dataset containing information on PM2.5 levels and AQI buckets in various cities.
Focused on the years 2015 to 2020 for comprehensive yearly analysis.

Analysis Methodology:
Employed statistical measures and visualizations to analyze PM2.5 levels across AQI buckets.
Utilized line charts, bar graphs, and other visualizations to represent yearly trends.

**3. Observations**

PM2.5 Distribution in Different AQI Buckets:

Visual representation of PM2.5 levels categorized into different AQI buckets for selected cities.
Yearly Trends:

Line charts illustrating the yearly trends of PM2.5 levels in each AQI bucket for specific cities.

**4. Managerial Implications or Insights**

Air Quality Management Strategies:
Identify cities with consistent poor air quality (elevated PM2.5 levels) and implement targeted measures to improve air quality.

Seasonal Variations:
Assess whether certain cities exhibit seasonal variations in PM2.5 levels and adjust strategies accordingly.

Public Awareness:
Communicate AQI bucket information to the public, fostering awareness and encouraging proactive measures during high pollution periods.

Regulatory Adjustments:
Evaluate the effectiveness of existing regulations and consider adjustments to address persistent air quality challenges.

**Conclusion:**

This analysis provides valuable insights into the distribution of PM2.5 levels across different AQI buckets in various cities from 2015 to 2020. The yearly trends offer a comprehensive understanding of air quality variations, guiding stakeholders in formulating effective air quality management strategies.

Sources-

Code reference-https://plotly.com/python/plotly-express/

Dataset-https://www.kaggle.com/datasets/rohanrao/air-quality-data-in-india/data